### PR TITLE
DM-21183: Update python coverage section of dev guide

### PR DESCRIPTION
**Pull request by parejkoj**
We are using pytest-cov+Coverage.py for automatic test coverage now:
these docs now reflect that, and provide some guidelines on how to use
the output.

### CODE REVIEW

File: stack/unit-test-coverage.rst [hunk 1]
```diff
@@ -33,22 +33,43 @@ code segments which have not been adequately tested and should then revise the
...
-.. _Code Coverage Analysis: http://www.bullseye.com/coverage.html
-.. _Minimum Acceptable Code Coverage: http://www.bullseye.com/minimum.html
+The LSST DM pytest suite runs `pytest-cov`_ to run `Coverage.py`_ when running tests via scons, producing a coverage report in the ``tests/.test/pytest-PACKAGE.xml-htmlcov`` directory.
+The exact path is printed at the end of the pytest secton of the scons output.
+Open the ``index.html`` file contained therein (for example, on a mac run ``open tests/.test/pytest-PACKAGE.xml-htmlcov/index.html`` to load it in your default browser) for a listing of the coverage percent for each file.
+Click on a filename in the list to get a line-by-line listing of what was and wasn't run.
 
-Using Coverage Analysis Tools
-=============================
+The "coverage" of a given file--shown on the summary page, and at the top of the individual file page--is the total lines of code excluding docstrings/comments/blank lines divided by the number of lines run.
```
Comment (timj) [inline stack/unit-test-coverage.rst:50]: Maybe add a couple of screenshots?

**Review by timj — APPROVED** [review]
Thanks for this. Looks okay.

Middleware packages have started to use CodeCov to give a coverage report inline on pull requests but since pipelines packages won't be able to do that any time soon mentioning it would probably confuse people.

File: stack/unit-test-coverage.rst [hunk 2]
```diff
@@ -33,22 +33,43 @@ code segments which have not been adequately tested and should then revise the
...
 
-A specific metric for lines of code executed and/or metric for branch
-conditionals executed is expected to be defined for Construction.
+Python
+------
 
-.. _Code Coverage Analysis: http://www.bullseye.com/coverage.html
-.. _Minimum Acceptable Code Coverage: http://www.bullseye.com/minimum.html
+The LSST DM pytest suite runs `pytest-cov`_ to run `Coverage.py`_ when running tests via scons, producing a coverage report in the ``tests/.test/pytest-PACKAGE.xml-htmlcov`` directory.
+The exact path is printed at the end of the pytest secton of the scons output.
```
Comment (ktlim) [inline stack/unit-test-coverage.rst:46]:
```suggestion
The exact path is printed at the end of the pytest section of the ``scons`` output.
```

File: stack/unit-test-coverage.rst [hunk 3]
```diff
@@ -33,22 +33,43 @@ code segments which have not been adequately tested and should then revise the
...
+Open the ``index.html`` file contained therein (for example, on a mac run ``open tests/.test/pytest-PACKAGE.xml-htmlcov/index.html`` to load it in your default browser) for a listing of the coverage percent for each file.
+Click on a filename in the list to get a line-by-line listing of what was and wasn't run.
 
-Using Coverage Analysis Tools
-=============================
+The "coverage" of a given file--shown on the summary page, and at the top of the individual file page--is the total lines of code excluding docstrings/comments/blank lines divided by the number of lines run.
+Red lines were not run by the tests you ran, yellow are "partially run" (e.g. an ``if``-statement condition that is only satisfied in one way).
+Imports, class and function declarations, and Configs are some examples of lines of code that get executed when a file is imported.
+Those can artificially inflate coverage statistics: a file with only 1-2 lines per function that is included in ``__init__.py`` might have >50% coverage, even though no tests actually run any of the funtions!
```
Comment (ktlim) [inline stack/unit-test-coverage.rst:53]:
```suggestion
Those can artificially inflate coverage statistics: a file with only 1-2 lines per function that is included in ``__init__.py`` might have >50% coverage, even though no tests actually run any of the functions!
```

File: stack/unit-test-coverage.rst [hunk 4]
```diff
@@ -179,3 +172,18 @@ Later when a Python unittester invokes an instrumented C++ routine, the
...
+
+Further reading
+===============
+
+Refer to `Code Coverage Analysis`_, by Steve Cornett, for a discussion of
+coverage metrics and to `Minimum Acceptable Code Coverage`_, also by Steve
+Cornett, for the companion discussion on determining 'good-enough' overall test
+coverage.
+
+A specific metric for lines of code executed and/or metric for branch
```
Comment (ktlim) [inline stack/unit-test-coverage.rst:184]: This never happened; might as well remove it.

**Review by ktlim — APPROVED** [review]
Thanks for this.  I agree with Tim, some screenshots might be nice.

**Comment by parejkoj** [comment]
I added a couple of screenshots with captions.

**Pull request merged**